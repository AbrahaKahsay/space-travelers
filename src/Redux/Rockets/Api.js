import { fetchRockets } from "./Rockets";
// get rockets from server
const url = 'https://api.spacexdata.com/v3/rockets';

const fetchRocketsFromServer = async (dispatch) => {
  const data = await fetch(url);
  const rockets = await data.json();
  dispatch(fetchRockets(rockets.map((rocket) => ({
    id: rocket.id,
    name: rocket.rocket_name,
    description: rocket.description,
    images: rocket.flickr_images,
  }))));
};

export default fetchRocketsFromServer;