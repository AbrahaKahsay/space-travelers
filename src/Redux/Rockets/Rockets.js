// Actions
const ROCKETS_FETCHED = 'ROCKETS_FETCHED';
const ROCKETS_RESERVED = 'ROCKETS_RESERVED';

// Actions creators
const fetchRockets = (payload) => ({
  type: ROCKETS_FETCHED,
  payload,
});
const reserveRocket = (id) => ({
  type: ROCKETS_RESERVED,
  id,
});

const initialState = [];
let newState;

// Reducers
const rocketsReducer = (state = initialState, action) => {
  switch (action.type) {
    case ROCKETS_FETCHED:
      return action.payload;
    case ROCKETS_RESERVED:
      newState = state.map((rocket) => {
        if(rocket.id !==action.id){
          return rokt;
        }
        else {
          return {...rocket, reserved:true};
        }
      })
    default:
      return state;
  }
};
export { fetchRockets };
export default rocketsReducer;