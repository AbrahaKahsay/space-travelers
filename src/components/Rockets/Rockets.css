.rocket-container {
  display: grid;
  grid-template-columns: 1.25fr 3fr;

  /* grid-template-rows: 1fr; */
  grid-column-gap: 0.5rem;
  grid-row-gap: 0.5rem;
  padding: 1rem;
}

.rocket-image {
  display: flex;
  height: 200px;
  width: 200px;
}

.r-image {
  height: 100%;
  width: 100%;
  object-fit: contain;
}

.name-descr-container {
  display: flex;
  flex-direction: column;
}

.res-btn {
  width: 150px;
  height: 30px;
  background-color: blue;
  border-style: none;
  color: white;
  border-radius: 4px;
}

.canc-btn {
  width: 150px;
  height: 30px;
  color: gray;
  border-radius: 4px;
}

.badge {
  width: 70px;
  height: 20px;
  background-color: rgb(70, 200, 240);
  border-style: none;
  color: white;
  margin-right: 5px;
  border-radius: 5px;
}