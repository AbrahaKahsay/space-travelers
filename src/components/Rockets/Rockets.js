import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import fetchRocketsFromServer from '../../Redux/Rockets/Api';

const Rockets = ()=>{
  const dispatch = useDispatch();
  const rockets = useSelector((state) => state.Rockets);
  useEffect(() => {
    if (!rockets.length) {
      dispatch(fetchRocketsFromServer);
    }
  }, []);
  return(
  <div>
    {rockets.map((rocket) => (
    <div key={rocket.id}>
      <img src={rocket.images[0]} alt="Rocket image" />
      <div>
        <h2 id={rocket.id}>{rocket.name}</h2>
        <p> {rocket.description}</p>
      </div>
    </div>
    ))}
   </div>
   );
}

export default Rockets;