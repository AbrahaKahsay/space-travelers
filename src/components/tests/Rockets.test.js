import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import Rockets from '../../components/Rockets/Rockets'
import store from '../../Redux/configureStore';
import rocketsReducer from '../../Redux/Rockets/Rockets';
import { fetchRocketsFromServer } from '../../Redux/Rockets/Rockets';

describe('Jest Snapshot testing suite', () => {
  it('Matches DOM Snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <Rockets />
        </Provider>
      </MemoryRouter>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should return an empty state', () => {
    const newState = rocketsReducer(undefined, {});
    expect(newState).toEqual([]);
  });

  it('should return content from the rocket API', () => {
    const rocketAPI = fetchRocketsFromServer();
    expect(rocketAPI).not.toEqual(null);
  });
})