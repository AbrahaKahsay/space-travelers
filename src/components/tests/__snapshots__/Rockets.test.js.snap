// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Snapshot testing suite Matches DOM Snapshot 1`] = `<div />`;
